# Lab 2 by Boying Li (boylil-3)

# Word2vec

## Dimension = 50

In [ ]:
# import modules & set up logging
import gensim, logging
import numpy as np
import help_functions as hf
import nltk

#@author: The first version of this code is the courtesy of Vadim Selyanik


logging.basicConfig(format='%(asctime)s : %(levelname)s : %(message)s', level=logging.INFO)
lemmatizer = nltk.WordNetLemmatizer() # create a lemmatizer

sentences = []
file = open("lemmatized.text", "r")

for line in file: # read the file and create list which contains all sentences found in the text
    sentences.append(line.split())
# train word2vec on the two sentences

dimension = 50 # parameter for Word2vec size of vectors for word embedding

# Words that occur more frequently than this threshold are randomly downsampled
threshold = 0.00055 # parameter for Word2vec

sum = 0.0

# sg=1 indicates using the skip-gram model(predict context word given target word), sg=0 means using cbow
model = gensim.models.Word2Vec(sentences, min_count=1, sample=threshold, sg=1, vector_size=dimension) # create model using Word2Ve with the given parameters

Note: gensim.models.Word2Vec() train for 5 epochs default if no epochs number specified.

In [ ]:
# model.vocab is replaced by model.wv
print(len(model.wv)) # check the length of the vocabulary which was formed by Word2Vec

70583


In [ ]:
nltk.download('wordnet')
nltk.download('omw-1.4')

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data] Downloading package omw-1.4 to /root/nltk_data...


True

In [ ]:
#The rest implements passing TOEFL tests
i = 0 #counter for TOEFL tests
number_of_tests = 80
text_file = open('new_toefl.txt', 'r')
right_answers = 0 # variable for correct answers
number_skipped_tests = 0 # some tests could be skipped if there are no corresponding words in the vocabulary extracted from the training corpus
while i < number_of_tests:
            line = text_file.readline() #read line in the file
            words = line.split() # extract words from the line
            try:
                words = [lemmatizer.lemmatize(lemmatizer.lemmatize(lemmatizer.lemmatize(word, 'v'), 'n'), 'a') for word in
                         words] # lemmatize words in the current test
                vectors = []
                if words[0] in model.wv: # check if there embedding for the query word
                    k = 1 #counter for loop iterating over 5 words in the test
                    vectors.append(model.wv[words[0]])
                    while k < 5:
                        if words[k] in model.wv: # if alternative has the embedding
                            vectors.append(model.wv[words[k]]) #assing the learned vector
                        else:
                            vectors.append(np.random.randn(dimension)) #assing random vector
                        k += 1
                    right_answers += hf.get_answer_mod(vectors) #find the closest vector and check if it is the correct answer

            except KeyError: # if there is no representation for the query vector than skip
                number_skipped_tests += 1
                print("skipped test: " + str(i) + "; Line: " + str(words))
            except IndexError:
                print(i)
                print(line)
                print(words)
                break
            i += 1
text_file.close()
sum += 100 * float(right_answers) / float(number_of_tests) #get the percentage

In [ ]:
print("Threshold ferq = "+ str(threshold)+" Percentage of correct answers: " + str(sum) + "%")

Threshold ferq = 0.00055 Percentage of correct answers: 70.0%


## Dimension = 5

In [ ]:
dimension = 5 # parameter for Word2vec size of vectors for word embedding

# Words that occur more frequently than this threshold are randomly downsampled
threshold = 0.00055 # parameter for Word2vec

sum = 0.0

# sg=1 indicates using the skip-gram model(predict context word given target word), sg=0 means using cbow
model = gensim.models.Word2Vec(sentences, min_count=1, sample=threshold, sg=1, vector_size=dimension) # create model using Word2Ve with the given parameters

In [ ]:
print(len(model.wv)) # check the length of the vocabulary which was formed by Word2Vec

70583


Note: The length of model.wv is determined by the number of unique words in your training data that meet the min_count threshold, so it remains the same with different vector_size(dimension)

In [ ]:
#The rest implements passing TOEFL tests
i = 0 #counter for TOEFL tests
number_of_tests = 80
text_file = open('new_toefl.txt', 'r')
right_answers = 0 # variable for correct answers
number_skipped_tests = 0 # some tests could be skipped if there are no corresponding words in the vocabulary extracted from the training corpus
while i < number_of_tests:
            line = text_file.readline() #read line in the file
            words = line.split() # extract words from the line
            try:
                words = [lemmatizer.lemmatize(lemmatizer.lemmatize(lemmatizer.lemmatize(word, 'v'), 'n'), 'a') for word in
                         words] # lemmatize words in the current test
                vectors = []
                if words[0] in model.wv: # check if there embedding for the query word
                    k = 1 #counter for loop iterating over 5 words in the test
                    vectors.append(model.wv[words[0]])
                    while k < 5:
                        if words[k] in model.wv: # if alternative has the embedding
                            vectors.append(model.wv[words[k]]) #assing the learned vector
                        else:
                            vectors.append(np.random.randn(dimension)) #assing random vector
                        k += 1
                    right_answers += hf.get_answer_mod(vectors) #find the closest vector and check if it is the correct answer

            except KeyError: # if there is no representation for the query vector than skip
                number_skipped_tests += 1
                print("skipped test: " + str(i) + "; Line: " + str(words))
            except IndexError:
                print(i)
                print(line)
                print(words)
                break
            i += 1
text_file.close()
sum += 100 * float(right_answers) / float(number_of_tests) #get the percentage

In [ ]:
print("Threshold ferq = "+ str(threshold)+" Percentage of correct answers: " + str(sum) + "%")

Threshold ferq = 0.00055 Percentage of correct answers: 37.5%


## Dimension = 500

In [ ]:
%%time
dimension = 500 # parameter for Word2vec size of vectors for word embedding

# Words that occur more frequently than this threshold are randomly downsampled
threshold = 0.00055 # parameter for Word2vec

sum = 0.0

# sg=1 indicates using the skip-gram model(predict context word given target word), sg=0 means using cbow
model = gensim.models.Word2Vec(sentences, min_count=1, sample=threshold, sg=1, vector_size=dimension) # create model using Word2Ve with the given parameters

CPU times: user 13min 59s, sys: 2.26 s, total: 14min 1s
Wall time: 8min 23s


In [ ]:
#The rest implements passing TOEFL tests
i = 0 #counter for TOEFL tests
number_of_tests = 80
text_file = open('new_toefl.txt', 'r')
right_answers = 0 # variable for correct answers
number_skipped_tests = 0 # some tests could be skipped if there are no corresponding words in the vocabulary extracted from the training corpus
while i < number_of_tests:
            line = text_file.readline() #read line in the file
            words = line.split() # extract words from the line
            try:
                words = [lemmatizer.lemmatize(lemmatizer.lemmatize(lemmatizer.lemmatize(word, 'v'), 'n'), 'a') for word in
                         words] # lemmatize words in the current test
                vectors = []
                if words[0] in model.wv: # check if there embedding for the query word
                    k = 1 #counter for loop iterating over 5 words in the test
                    vectors.append(model.wv[words[0]])
                    while k < 5:
                        if words[k] in model.wv: # if alternative has the embedding
                            vectors.append(model.wv[words[k]]) #assing the learned vector
                        else:
                            vectors.append(np.random.randn(dimension)) #assing random vector
                        k += 1
                    right_answers += hf.get_answer_mod(vectors) #find the closest vector and check if it is the correct answer

            except KeyError: # if there is no representation for the query vector than skip
                number_skipped_tests += 1
                print("skipped test: " + str(i) + "; Line: " + str(words))
            except IndexError:
                print(i)
                print(line)
                print(words)
                break
            i += 1
text_file.close()
sum += 100 * float(right_answers) / float(number_of_tests) #get the percentage
print("Threshold ferq = "+ str(threshold)+" Percentage of correct answers: " + str(sum) + "%")

Threshold ferq = 0.00055 Percentage of correct answers: 71.25%


Summary: I have tried dimension of 5(accuracy=37.5%), 50(accuracy=70%) and 500(accuracy=71.25%). With higher dimension, the accuracy is higher because higher dimensional vectors can capture more detailed and nuanced relationships between words, but it also uses more memory and computational power and longer training time. Dimension 50 improves the accuracy significantly compared to dimension 5, while continue increase the dimension to 500 does not increase the accuracy that much. So 50 dimensional vectors are less detailed compared to 500 dimensional vectors but 50 dimensional can still capture the most important relationships.